include/maliput/routing/route.h

```diff
@@ -0,0 +1,122 @@
+#include "maliput/common/maliput_copyable.h"
+#include "maliput/common/maliput_throw.h"
+#include "maliput/routing/route_phase.h"
+
+namespace maliput {
+namespace routing {
+
+/// Describes the sequence of paths to go from one RoadPosition to another.
+/// It is sequenced by RoutePhases which contain default and adjacent
+/// maliput::api::LaneSRanges which conduce the agent from the start to the end
+/// road position.
+///
+/// Agents are expected to use the Router to obtain a Route. Once in the Route,
+/// they can iterate through the RoutePhases or find by a specific INERTIAL or
+/// LANE Frame coordinate the RoutePhase and start driving from there towards
+/// the end goal.
+///
+/// The first RoutePhase start road position identifies the beginning of the
+/// Route. The last RoutePhase end road position identifies the ending of the
+/// Route. RoutePhases in the sequence are connected end to end by the same
+/// INERTIAL-Frame coordinate whose the corresponding LANE-Frame coordinate
+/// depends on which maliput::api::LaneSRange they contain.
+class Route final {
+ public:
+  MALIPUT_DEFAULT_COPY_AND_MOVE_AND_ASSIGN(Route);
+  Route() = delete;
+
+  /// Constructs a Route.
+  ///
+  /// @param route_phases The sequence of RoutePhases. It must not be nullptr.
+  /// RoutePhases must be connected end to end.
+  /// @param road_network The maliput::api::RoadNetwork pointer. It must not be
+  /// nullptr.
+  /// @throws maliput::common::assertion_error When @p route_phases is empty.
+  /// @throws maliput::common::assertion_error When @p route_phases contains
+  /// RoutePhases that are not connected end to end.
+  /// @throws maliput::common::assertion_error When @p road_networ is nullptr.
+  Route(const std::vector<RoutePhases>& route_phases, const maliput::api::RoadNetwork* road_network) :
+      route_phases_(route_phases), road_network_(road_network) {
+    MALIPUT_THROW_UNLESS(!route_phases_.empty());
+    MALIPUT_THROW_UNLESS(road_network_ != nullptr);
+    /// TODO(#453): Validate end to end connection of the RoutePhases.
+  }
+
+  /// @return The number of RoutePhases.
+  int Size() const { static_cast<int>(route_phases_.size()); }
+
+  /// Indexes the RoutePhases.
+  ///
+  /// @param index The index of the RoutePhase. It must be not be negative and
+  /// less than `size()`.
+  /// @return The RoutePhase at @p index.
+  /// @throws std::out_of_range When @p index is negative or >= `size()`.
+  const RoutePhase& Get(int index) const { return route_phases_.at(index); }
+
+  /// @return The start maliput::api::RoadPosition of this Route.
+  const maliput::api::RoadPosition& StartRoadPosition() const {
+    return route_phases_.front().StartRoadPosition();
+  }
+
+  /// @return The ebd maliput::api::RoadPosition of this Route.
+  const maliput::api::RoadPosition& EndRoadPosition() const {
+    return route_phases_.back().EndRoadPosition();
+  }
+
+  /// Finds the RoutePhase where @p inertial_position falls into.
```

Comment:
I think it makes sense to be consistent and use the same criteria as in `RoadGeometry::ToRoadPosition()`. However, I also think it is not the intended use because while querying the `Route`, being able to distinguish if a coordinate falls into or not the `Route` allows the agent behavior to trigger a new plan or to fail (when there is no recovery path).

If we decided to answer a `RoutePhase` that minimizes the distance to the `inertial_position` we would be able to query any coordinate, and still map a `RoutePhase`. Would it be useful for the agent behavior?

Reply:
Here's a situation where this might be useful: Suppose an agent is an off-road-capable vehicle like a Toyota TRD 4Runner. On occasion, it might adventure off-road but later wish to get back to a well-traveled on-road path. In such an application, it might wish to get the closest `RoutePhase` even if doesn't overlap with the agent's current position. In general, there are many reasons why an agent's inertial position may end up off road, and allowing for them to still query for a route seems potentially useful.

Reply:
I agree that we need to provide the closest route position in this case. Taking a simpler example of a controller we're trying to train, the car will fall off the pavement here and there so we want a notion of when to get back it. Furthermore the decision to re-route may be made by an agent based on the distance from a point on the route. "I am far enough away, better re-route"

Reply:
Alright! I'll work on the changes.